### PR TITLE
Add miopen.lds_barrier operator with translation to gpu.lds_barrier to rocdl.lds_barrier

### PR DESCRIPTION
**Pull request by sjw36**
Merged from miopen-dialect-index-diff-map branch

This feature allows generation of a barrier specific to LDS stores. It generates an asm implementation in the LLVMIR dialect with these instructions:

> s_waitcnt lgkmcnt(0)
> s_barrier

MLIR#120: https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/120

### CODE REVIEW

File: mlir/test/Target/rocdl.mlir [hunk 1]
```diff
@@ -1,4 +1,6 @@
 // RUN: mlir-translate -mlir-to-rocdlir %s | FileCheck %s
+// CHECK: target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-ni:7"
```
Comment (whchung) [inline mlir/test/Target/rocdl.mlir:2]: we probably don't need these checks around datalayout and triple.

I'll be submitting another PR where we deliberately change width of pointers in generic address space to 32-bits so we tame the width of MLIR `index` type be the same as `i32` so avoid unnecessary integer upcast/downcast while doing GEP/load/store.

**Review by whchung — COMMENTED** [review]
thanks! one minor change in the unit test and this PR looks good to go.

**Comment by whchung** [comment]
jenkins: retest this please.

@sjw36 let's disable flaky e2e tests should it persist.

**Comment by sjw36** [comment]
Yes, it also fails this way on TOT.

**Comment by whchung** [comment]
@sjw36 got it. Let’s strike out those flaky fp16 tests.

**Review by whchung — CHANGES_REQUESTED** [review]
@sjw36 could we just strike out those flaky test configs, instead of killing the whole set?

**Comment by sjw36** [comment]
@whchung ok, let me narrow down.

**Review by whchung — APPROVED** [review]

**Pull request merged**